### PR TITLE
BLUE-2644-label-for

### PR DESCRIPTION
**Pull request by richardicke**

### CODE REVIEW

File: plone/recipe/codeanalysis/a11y.py [hunk 1]
```diff
@@ -146,13 +148,25 @@ def missing_button_content(context):
...
 
+@Context.add_check
+def missing_for(context):
+    for label in context.node.xpath('//xhtml:label|//label', namespaces=NSMAP):
+        label_for = attribute(label, 'for')
+        if label_for is None:
+            context.report(
+                label,
+                'The <label> element needs to be explicitly associated with a '
+                'form control through the use of the for attribute.')
```
Comment (redrohX) [inline plone/recipe/codeanalysis/a11y.py]: Perhaps you should also mention the `for` value should be the same as the value of the `id` attribute of the form control.

Comment (richardicke) [reply plone/recipe/codeanalysis/a11y.py]: ok

File: plone/recipe/codeanalysis/tests/test_a11y_lint.py [hunk 2]
```diff
@@ -298,6 +307,12 @@ def test_button_has_tal_content_child_node(self):
         with OutputCapture():
             self.assertTrue(A11yLint(self.options).run())
 
+    def test_label_has_for_attribute(self):
```
Comment (janwijbrand) [inline plone/recipe/codeanalysis/tests/test_a11y_lint.py]: I think it is useful to also have the positive check in a test, so, when there indeed is a `for` attribute present.

Comment (richardicke) [reply plone/recipe/codeanalysis/tests/test_a11y_lint.py]: ok

File: plone/recipe/codeanalysis/tests/test_a11y_lint.py [hunk 3]
```diff
@@ -29,6 +29,16 @@
 </html>
 """
 
+LINK_ROLE_BUTTON = """\
+<html
+  xmlns="http://www.w3.org/1999/xhtml"
+  xmlns:tal="http://xml.zope.org/namespaces/tal">
+  <body>
+    <a role="button" href="#">Link</a>
```
[the remark this answers is not in the document]
Comment (redrohX) [reply plone/recipe/codeanalysis/tests/test_a11y_lint.py]: No, without the `href` it's impossible to reach the button by keyboard navigation.

File: plone/recipe/codeanalysis/a11y.py [hunk 4]
```diff
@@ -76,6 +76,10 @@ def run(self):
 def missing_href(context):
     for link in context.node.xpath('//xhtml:a|//a', namespaces=NSMAP):
         href = attribute(link, 'href')
+        if attribute(link, 'role') == 'button':
+            continue
```
Comment (gyst) [inline plone/recipe/codeanalysis/a11y.py]: Niels says: 
> without the href it's impossible to reach the button by keyboard navigation

However, this code change makes exactly that possible: a link with role button without href.

Either we're miscommunicating, or this is off.

**Pull request merged**